src/devices/sound/discrete.cpp

```diff
@@ -204,74 +209,78 @@ inline void discrete_task::step_nodes()
 		}
 	}
 
-	/* buffer the outputs */
+	// buffer the outputs
+	for (output_buffer &outbuf : m_buffers)
+		*outbuf.ptr.load(std::memory_order_relaxed) = *outbuf.source;
+	std::atomic_thread_fence(std::memory_order_release);
 	for (output_buffer &outbuf : m_buffers)
-		*outbuf.ptr++ = *outbuf.source;
+		outbuf.ptr.store(outbuf.ptr.load(std::memory_order_relaxed) + 1, std::memory_order_relaxed);
 }
 
 void *discrete_task::task_callback(void *param, int threadid)
 {
 	const auto &list = *reinterpret_cast<const discrete_sound_device::task_list_t *>(param);
-	do
+	while (true)
 	{
 		for (const auto &task : list)
 		{
-			/* try to lock */
+			// try to lock
 			if (task->lock_threadid(threadid))
 			{
 				if (!task->process())
 					return nullptr;
 				task->unlock();
 			}
 		}
-	} while (1);
+	}
 
 	return nullptr;
 }
 
-bool discrete_task::process()
+inline bool discrete_task::process()
 {
-	int samples = std::min(int(m_samples), MAX_SAMPLES_PER_TASK_SLICE);
+	int samples = std::min(m_samples, MAX_SAMPLES_PER_TASK_SLICE);
 
-	/* check dependencies */
+	// check dependencies
 	for (input_buffer &sn : source_list)
 	{
-		int avail = sn.linked_outbuf->ptr - sn.ptr;
+		const int avail = sn.linked_outbuf->ptr.load(std::memory_order_relaxed) - sn.ptr;
 		if (avail < 0)
 			throw emu_fatalerror("discrete_task::process: available samples are negative");
 		if (avail < samples)
 			samples = avail;
 	}
+	std::atomic_thread_fence(std::memory_order_acquire);
```

Comment:
We only need a single acquire barrier here between determining the minimum number of samples available across all inputs and consuming the samples.

Since `output_buffer::ptr` is atomic, we get a guarantee that there’s a single load here and we won’t get an inconsistent value.